sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1372,6 +1376,108 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
   }
 }
 
+/**
+ * Rewriting join condition to conjunctive normal form expression so that we can push
+ * more predicate.
+ */
+object PushPredicateThroughJoinByCNF extends Rule[LogicalPlan] with PredicateHelper {
+  /**
+   * Rewrite pattern:
+   * 1. (a && b) || c --> (a || c) && (b || c)
+   * 2. a || (b && c) --> (a || b) && (a || c)
+   *
+   * To avoid generating too many predicates, we first group the filter columns from the same table.
+   */
+  private def toCNF(condition: Expression, depth: Int = 0): Expression = {
+    if (depth < SQLConf.get.maxRewritingCNFDepth) {
+      condition match {
+        case or @ Or(left: And, right: And) =>
+          val lhs = splitConjunctivePredicates(left).groupBy(_.references.map(_.qualifier))
+          val rhs = splitConjunctivePredicates(right).groupBy(_.references.map(_.qualifier))
+          if (lhs.size > 1) {
+            lhs.values.map(_.reduceLeft(And)).map { c =>
+              toCNF(Or(toCNF(c, depth + 1), toCNF(right, depth + 1)), depth + 1)
```

Comment:
I don't really understand this. At least I can't see how is it related to `(a && b) || c --> (a || c) && (b || c)`

Reply:
This is a special case of **`(a && b) || c`**. E.g.: `(a && b) || (c && d)`. The rewriting steps are:
`(a && b) || (c && d)` --> `(a || (c && d)) && (b || (c && d))` --> `(a || c) && (a || d) && (b || c) && (b && d)`. 

TPC-DS Q13:
![image](https://user-images.githubusercontent.com/5399861/83621501-5154bb00-a5c1-11ea-9d4a-85483463a208.png)

Reply:
I added a complete step to PR description to explain how it works.